lib/portlayer/metrics/container.go

```diff
@@ -224,117 +215,200 @@ func (cc *collectorVM) collect(currentSubs map[types.ManagedObjectReference]*sub
+	_, exists := cc.subs[contain]
 	if !exists {
-		// create a new publisher with 100ms timeout and no buffer
-		containerSub.publisher = pubsub.NewPublisher(100*time.Millisecond, 0)
-		cc.subs[containerSub.ref] = containerSub
-		sub = containerSub
+		sub, err := newContainerSubscription(contain)
+		if err != nil {
+			return nil, err
+		}
+		cc.subs[contain] = sub
+	}
+
+	// get a subscriber channel
+	ch := cc.subs[contain].Channel()
+
+	// This is first subscription so start collection
+	if subscriptionCount == 0 {
+		cc.Start()
 	}
-	// subscribe to this publisher -- a publisher can have multiple subscribers
-	ch := sub.publisher.Subscribe()
+
 	return ch, nil
 }
 
-// Unsubscribe unsubscribes from the container metrics publisher
-func (cc *collectorVM) Unsubscribe(contain interface{}, ch chan interface{}) {
+// Unsubscribe from a container metric subscription.  The subscriber channel will
+// be evicted and when no subscribers remain the subscription will be removed.
+func (cc *vmCollector) Unsubscribe(contain interface{}, ch chan interface{}) {
 	defer trace.End(trace.Begin(""))
-	containerSub, err := containerSubscriber(contain)
-	if err != nil {
-		return
-	}
 	cc.mu.Lock()
 	defer cc.mu.Unlock()
-	sub, exists := cc.subs[containerSub.ref]
+
+	sub, exists := cc.subs[contain]
 	if exists {
-		sub.publisher.Evict(ch)
-		if sub.publisher.Len() == 0 {
-			delete(cc.subs, sub.ref)
+		// remove the communication channel
+		sub.Evict(ch)
+		// do we have any subscribers to this subscription
+		if sub.Publishers() == 0 {
+			log.Debugf("Deleting container(%s) subscription", cc.subs[contain].containerID())
+			delete(cc.subs, contain)
 		}
 	}
+
+	// no subscriptions, so stop the collection
+	if len(cc.subs) == 0 {
+		cc.Stop()
+	}
 }
 
-// Subscribers returns the current subscribers
-func (cc *collectorVM) Subscribers() map[types.ManagedObjectReference]*subscriber {
-	current := make(map[types.ManagedObjectReference]*subscriber)
+// Subscribers returns the current subscriptions by ManagedObject Reference
+func (cc *vmCollector) Subscribers() map[types.ManagedObjectReference]*subscription {
+	current := make(map[types.ManagedObjectReference]*subscription)
 	cc.mu.Lock()
 	defer cc.mu.Unlock()
-	for ref := range cc.subs {
-		current[ref] = cc.subs[ref]
+	for i := range cc.subs {
+		current[cc.subs[i].c.VMReference()] = cc.subs[i]
 	}
 	return current
 }
 
-// SubscribeCount will return the current number of subscribers
-func (cc *collectorVM) SubscriberCount() int {
-	cc.mu.Lock()
-	defer cc.mu.Unlock()
-	return len(cc.subs)
+// instanceID coverts the ID or Key of a metric from a string to int.
+func instanceID(containerID string, instance string) (int, error) {
+	converted, err := strconv.Atoi(instance)
+	if err != nil {
+		// I don't expect this to ever happen, but if it does log and don't publish
+		log.Errorf("metrics failed to convert the container(%s) device id to an int - value(%#v): %s", containerID, instance, err)
+		return converted, err
```

Comment:
I am not familiar enough with golang Atoi, but since you move this here and out of the switch statement above - does Atoi at least populate converted with 0/-1 or something sensible so that we are guaranteed we are not returning some totally invalid value?